### ISSUE TITLE
Search for matching while should be top-down not bottom-up

### ISSUE THREAD

**Issue by matter123** [microsoft/vscode-textmate#110]
Given the following grammar.
```json
"patterns": [
	{
		"match": "//.*",
		"name": "comment"
	},
	{
		"name": "meta.outer-pattern",
		"begin": "if",
		"end": "^\\s*+endif",
		"patterns": [
			{
				"name": "meta.inner-pattern",
				"begin": "\\G",
				"while": "^\\s*+(?!endif)",
				"patterns": [
					{
						"name": "keyword.control.test",
						"match": "test"
					},
					{
						"include": "$self"
					}
				]
			}
		]
	}
]
```

And the following test file (comments are of expected highlighting)

```
// outside of if, no highlighting
test
if
// inside of if, highlighting
test
endif

// scopes close properly, no highlighting
test

if
if
// inside of nested if, highlighting
test
endif
// inside of outer if, highlighting
test
endif

// scopes close properly, no highlighting
test
```

The 5th and 6th `test` are highlighted incorrectly as compared to TextMate.

The 5th `test` is not highlighted because vscode-textmate searches from the bottom up and closes both the outer and inner begin/while rule.

The 6th `test` is highlighted because the unconsumed final `endif` matches /if/.

Textmate has the 5th `test` highlighted and the 6th not.
<img width="488" alt="Screen Shot 2019-08-21 at 20 19 19" src="https://user-images.githubusercontent.com/714007/63483403-209bfe80-c451-11e9-94ed-ce054c7bd2a2.png">

**Comment by msftrncs** [comment]
Analyzing this I am pretty sure the correct way to check the rules is from bottom to top.

While rules are specifically for constructs such as MarkDown Quote Block, which becomes nested many times.

> nest 1
>
> > nest 2 (while rule tests # 1 then # 2 on each line.)
>
> nest 1

Now, with that said, I think TextMate might be detecting that the same rule is nested, and it only pulls the stack back to the point of the top most instance.  This would also work for MarkDown just fine.  However, this is more complicated than it sounds.  If there are any other while rules between the highest instance of the tested rule and the furthest down one, those have to be tested too.

So, flip the stack of rules to a new pile (like currently done), starting with the bottom most while rule (now on the top of the while stack), test if it matches, if not, search the remaining flipped stack for the furthest down instance of the same rule, and if found, pull the main stack back to this point and keep only the top of the flipped stack above the furthest instance, then resume testing the flipped stack, else (if no further instance found) pull the stack back to this point and return.

I think I have a working demonstration of this concept now.

**Comment by msftrncs** [comment]
For more history on the topic, see previous issue #25.  

@mjbvz, do you have any input on this issue or the PR #111?

referring back to #25:
> 1. On a new line, check all while conditions from bottom of the stack to the top.
> 2. If any of them fails, then pop everything after that while rule off the stack, along with the failed rule.

By nothing more than analytical thinking, I am suspecting its actually:
1. On a new line, check all while conditions from bottom of the stack to the top.
2. For each while condition that fails:
   1. Search stack from top to bottom for another instance of the same while condition.
   2. Pop everything after and including the first (highest) instance found.
   3. Continue checking remaining while conditions.

**Comment by msftrncs** [comment]
I've had a revelation of thought while trying to understand why VS Code displays Markdown quotes in preview differently than in the editor.

Example:
```markdown
> one nest.
>> second nest, new paragraph.
> continues paragraph of second nest.
also continues paragraph of second nest.
```

Actually (via GitHub):
> one nest.
> > second nest, new paragraph.
> continues paragraph of second nest.
also continues paragraph of second nest.

VS Code Editor highlight:
![image](https://user-images.githubusercontent.com/26179051/63668222-1ef97000-c79c-11e9-9727-72d862708081.png)

From VS Code Preview:
![image](https://user-images.githubusercontent.com/26179051/63668189-01c4a180-c79c-11e9-96f3-cb1ee3cf0a2a.png)

I believe VS Code's Markdown grammar is wrong.  The while rule is expecting to find additional lines starting with `>` but that is not how the previewer works.  Using the flawed Markdown grammar as a basis for the correct processing order of `while` conditions copies the flaw to that processing.  I believe the correct test is, at a minimum, a non-blank line. but possibly also tests for other kinds of markup, or precisely the same as the `while` clause of `list_paragraph`.    This is the correct `while` condition for `blockquote`: ```"while": "(^|\\G)([ ]{0,3}(>) ?|(?!\\s*$|#|[ ]{0,3}([-*_>][ ]{2,}){3,}[ \\t]*$\\n?|[ ]{0,3}[*+->]|[ ]{0,3}[0-9]+\\.))"```.  With this change, I get the correct scoping, and it would appear to me that I would get the correct scoping if the `while` conditions were tested in the other order.

I think there is a chance that the `while` conditions should be tested top down, not bottom up.

I have closed PR #111 for the time being, at least until I can put more thought/testing in to whether it is correct.

**Comment by mjbvz** [comment]
I believe the current behavior is the expected design. Since you can workaround the original problem , let me know if you still think we should revisit this
